clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -6901,7 +6901,7 @@ CodeGenModule::getLLVMLinkageForDeclarator(const DeclaratorDecl *D,
   // have this, linkonce_odr suffices. If -fno-sycl-rdc is passed, we know there
   // is only one translation unit and can so mark them internal.
   if (getLangOpts().SYCLIsDevice && !D->hasAttr<DeviceKernelAttr>() &&
-      !D->hasAttr<SYCLDeviceAttr>() &&
+      !(D->hasAttr<SYCLDeviceAttr>() || D->hasAttr<SYCLExternalAttr>()) &&
```

Comment:
I think this would be slightly clearer without the disjunction but I'm ok with this if that is your preference.
```suggestion
      !D->hasAttr<SYCLDeviceAttr>() && !D->hasAttr<SYCLExternalAttr>() &&
```